reader.go

```diff
@@ -1203,7 +1203,11 @@ func (r *Reader) FetchMessage(ctx context.Context) (Message, error) {
 					m.error = io.ErrUnexpectedEOF
 				}
 
-				return m.message, m.error
+				msg, err := m.message.Decode()
+				if err != nil {
+					return m.message, err
+				}
+				return msg, m.error
```

Comment:
I think the behavior of this code is being changed if an error occurred before decoding, it used to return `m.error` but now it will likely return the decoder's own error instead.

It seems to me that the correct approach would be to check for `m.error != nil` before decoding the message, and if non-nil, return the error without attempting to decode the message.